### DISCUSSION TITLE
Custom Kernels/Operators for PlaidML

### DISCUSSION THREAD

**Discussion by artyom-beilis** [plaidml/plaidml#1870] [General]
Is there any way to add custom operators. I have relatively good implementations of convolutions and gemm for AMD and NVidia platforms: https://github.com/artyom-beilis/dlprimitives it is opencl "cudnn"

Since plaidml already works on OpenCL I want to "speed it up" using custom implementations. I 

Where can I start - what is the point to look into?

I started doing it on Caffe-OpenCL - but I'm familiar with it. Now were do I need look to inside PladiML?

**Comment by artyom-beilis** [comment]
Should I look into direction of `plaidml/bridge`?

Anybody?

**Comment by brianretford** [comment]
Speed up what exactly? I'm a. bit confused as to what you're trying to accomplish

**Reply by artyom-beilis** [reply]
I explain. Today plaidml performance is far from being close to performance of cudnn/miopen based frameworks, for example these are my benchmarks for resnet18, batch=16, train times in ms on gtx1080:

|Framework | time ms |
|-----------|---------|
|pytorch   | 38.359  |
|tf2       | 68.63 |
|dlprim    | 70.496 |
|caffe/ocl | 165.449 |
|plaidml   | 235.163 |

My project [dlprimitives](https://github.com/artyom-beilis/dlprimitives) is both micro-framework but also it provides cudnn/miopen like [core component](http://dlprimitives.org/docs/namespacedlprim_1_1core.html) but for opencl and it works on nvidia, amd, intel. At least for AMD and nVidia it gives very decent performance.

Since most of time DL framework spends in convolutions and gemms (that I have relatively good implementation for them) I wanted to add them to plaidml as "accelerator".

**Comment by artyom-beilis** [comment]
So?
